File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/RotateTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class RotateTest {

	@Test
	void test() {
		var array = new JsonArray();
		array.add("I");
		array.add("am");
		array.add("using");
		array.add("eclipse");
		array.add("to");
		array.add("test");
		array.add("the");
		array.add("changes");
		array.add("with");
		array.add("test");
		array.add("Driven");
		array.add("developement");

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

		fep.setArguments(Map.of("source", array, "rotateDistance", new JsonPrimitive(4))).setContext(Map.of())
				.setSteps(Map.of());

		var res = new JsonArray();

		res.add("to");
		res.add("test");
		res.add("the");
		res.add("changes");
		res.add("with");
		res.add("test");
		res.add("Driven");
		res.add("developement");

		res.add("I");
		res.add("am");
		res.add("using");
		res.add("eclipse");

		Rotate rotate = new Rotate();
		rotate.execute(fep).block();

		assertEquals(res, array);

		ReactiveFunctionExecutionParameters fep1 = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

		fep1.setArguments(Map.of("source", array, "rotateDistance", new JsonPrimitive(0))).setContext(Map.of())
				.setSteps(Map.of());

		StepVerifier.create(rotate.execute(fep1)).expectError(KIRuntimeException.class).verify();
	}

	@Test
	void test2() {
		var array = new JsonArray();
		array.add("I");
		array.add("am");
		array.add("using");
		array.add("eclipse");
		array.add("to");
		array.add("test");
		array.add("the");
		array.add("changes");
		array.add("with");
		array.add("test");
		array.add("Driven");
		array.add("developement");

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

		fep.setArguments(Map.of("source", array, "rotateDistance", new JsonPrimitive(4))).setContext(Map.of())
				.setSteps(Map.of());

		var res = new JsonArray();

		res.add("to");
		res.add("test");
		res.add("the");
		res.add("changes");
		res.add("with");
		res.add("test");
		res.add("Driven");
		res.add("developement");

		res.add("I");
		res.add("am");
		res.add("using");
		res.add("eclipse");

		Rotate rotate = new Rotate();
		rotate.execute(fep).block();

		assertEquals(res, array);

		ReactiveFunctionExecutionParameters fep1 = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

		fep1.setArguments(Map.of("source", array, "rotateDistance", new JsonPrimitive(-2))).setContext(Map.of())
				.setSteps(Map.of());

		StepVerifier.create(rotate.execute(fep1)).expectError(KIRuntimeException.class).verify();
	}

}

File: kirun-js/__tests__/engine/function/system/array/RotateTest.ts
import { Rotate } from '../../../../../src/engine/function/system/array/Rotate';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';
import { KIRunFunctionRepository, KIRunSchemaRepository } from '../../../../../src';

let rotate: Rotate = new Rotate();

test('Rotate test1 ', async () => {
    let array: any[] = [];
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Rotate.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
                [Rotate.PARAMETER_ROTATE_LENGTH.getParameterName(), 16],
            ]),
        )
        .setContext(new Map([]))
        .setSteps(new Map([]));

    let res: any[] = [];
    res.push('to');
    res.push('test');
    res.push('the');
    res.push('changes');
    res.push('with');
    res.push('test');
    res.push('Driven');
    res.push('developement');

    res.push('I');
    res.push('am');
    res.push('using');
    res.push('eclipse');

    expect((await rotate.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(
        res,
    );
});

test('rotate  test 2', async () => {
    let src: any[] = [];

    src.push('I');
    src.push('am');
    src.push('using');
    src.push('eclipse');
    src.push('to');
    src.push('test');
    src.push('the');
    src.push('changes');
    src.push('with');
    src.push('test');
    src.push('Driven');
    src.push('developement');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([
                [Rotate.PARAMETER_ARRAY_SOURCE.getParameterName(), src],
                [Rotate.PARAMETER_ROTATE_LENGTH.getParameterName(), src.length - 1],
            ]),
        )
        .setContext(new Map([]))
        .setSteps(new Map([]));

    let res: any[] = [];

    res.push('developement');
    res.push('I');
    res.push('am');
    res.push('using');
    res.push('eclipse');
    res.push('to');
    res.push('test');
    res.push('the');
    res.push('changes');
    res.push('with');
    res.push('test');
    res.push('Driven');

    expect((await rotate.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(
        res,
    );
});

test('rotate test 3', async () => {
    let array: any[] = [];
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    )
        .setArguments(
            new Map<string, any>([[Rotate.PARAMETER_ARRAY_SOURCE.getParameterName(), array]]),
        )
        .setContext(new Map([]))
        .setSteps(new Map([]));

    let res: any[] = [];
    res.push('am');
    res.push('using');
    res.push('eclipse');
    res.push('to');
    res.push('test');
    res.push('the');
    res.push('changes');
    res.push('with');
    res.push('test');
    res.push('Driven');
    res.push('developement');

    res.push('I');

    expect((await rotate.execute(fep)).allResults()[0].getResult().get('result')).toStrictEqual(
        res,
    );
});
